src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -206,5 +290,120 @@ public Task UnloadAsync()
 
             return Task.CompletedTask;
         }
+
+        // NuGet calls this method to wait project to nominate restoring.
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project nominate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to nominate a restore (DT build failed)
+        public Task WhenNominated(CancellationToken cancellationToken)
+        {
+            lock (SyncObject)
+            {
+                if (!CheckIfHasPendingNomination() || cancellationToken.IsCancellationRequested)
+                {
+                    return Task.CompletedTask;
+                }
+
+                _whenNominatedTask ??= new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            }
+
+            var continuationTask = _whenNominatedTask.Task.ContinueWith(t =>
+            {
```

Comment:
sorry, not continuation of this task, but this.SourceBlock.Completion, basically, when something wrong in the project system which failed the dataflow block, the final block (SourceBlock) will be completed with a failure exception. Under that condition, the project no longer processes any data (and will never trigger any new nomination), so the logic here is that we need notify anyone still waiting.  It need be done only once because it will complete the final _whenNormnatedTask if it is not null. (another place to put this code is in the InitializeAsync(), but it would be a waste, if WhenNominated has never been called for the project.

Also, you need check _whenNominatedTask is not null, because we need clear it out when it is completed. Otherwise WhenNominated would only work once.

Reply:
fixed in latest commit